File: example/src/calculator.js
/**
 * Calculator - 被测试的代码
 *
 * 这是一个简单的计算器模块，用于演示 TiaCC 的工作原理。
 */

export class Calculator {
  constructor() {
    this.history = [];
  }

  // 基础运算
  add(a, b) {
    const result = a + b;
    this.history.push({ op: 'add', a, b, result });
    return result;
  }

  subtract(a, b) {
    const result = a - b;
    this.history.push({ op: 'subtract', a, b, result });
    return result;
  }

  multiply(a, b) {
    const result = a * b;
    this.history.push({ op: 'multiply', a, b, result });
    return result;
  }

  divide(a, b) {
    if (b === 0) {
      throw new Error('Division by zero');
    }
    const result = a / b;
    this.history.push({ op: 'divide', a, b, result });
    return result;
  }

  // 高级运算
  power(base, exponent) {
    const result = Math.pow(base, exponent);
    this.history.push({ op: 'power', base, exponent, result });
    return result;
  }

  sqrt(n) {
    if (n < 0) {
      throw new Error('Cannot compute square root of negative number');
    }
    const result = Math.sqrt(n);
    this.history.push({ op: 'sqrt', n, result });
    return result;
  }

  // 历史记录
  getHistory() {
    return [...this.history];
  }

  clearHistory() {
    this.history = [];
  }

  getLastResult() {
    if (this.history.length === 0) {
      return null;
    }
    return this.history[this.history.length - 1].result;
  }
}

export default Calculator;

File: tests/e2e/cpp-project/src/calculator.cpp
#include "calculator.h"
#include <cmath>
#include <stdexcept>

// 基础四则运算

double Calculator::add(double a, double b) {
    return a + b;
}

double Calculator::subtract(double a, double b) {
    return a - b;
}

double Calculator::multiply(double a, double b) {
    return a * b;
}

double Calculator::divide(double a, double b) {
    if (b == 0.0) {
        throw std::invalid_argument("Division by zero");
    }
    return a / b;
}

// 高级运算

double Calculator::power(double base, int exponent) {
    if (exponent == 0) {
        return 1.0;
    }

    double result = 1.0;
    int absExp = exponent > 0 ? exponent : -exponent;

    for (int i = 0; i < absExp; ++i) {
        result *= base;
    }

    if (exponent < 0) {
        return 1.0 / result;
    }

    return result;
}

double Calculator::sqrt(double x) {
    if (x < 0) {
        throw std::invalid_argument("Cannot compute square root of negative number");
    }
    return std::sqrt(x);
}

double Calculator::abs(double x) {
    if (x < 0) {
        return -x;
    }
    return x;
}

// 累加器功能

void Calculator::reset() {
    accumulator_ = 0.0;
}

void Calculator::accumulate(double value) {
    accumulator_ += value;
}

double Calculator::getAccumulator() const {
    return accumulator_;
}
